smaps/src/main/java/com/google/smaps/Histogram.java

```diff
@@ -0,0 +1,136 @@
+    // Get numbers from form in interactive-histogram.html.
+    String lower = request.getParameter("lower-bound");
+    String upper = request.getParameter("upper-bound");
+    postFired = true;
+
+    // In case user entered a number with .0 as the decimal, cast to long to drop the decimal,
+    // otherwise just convert the String to a Long.
+    if (lower.contains(".")) {
+      double lowerDecimal = Double.parseDouble(lower);
+      lowerBound = (long) lowerDecimal;
+    } else {
+      lowerBound = Long.parseLong(lower);
+    }
+    if (upper.contains(".")) {
+      double upperDecimal = Double.parseDouble(upper);
+      upperBound = (long) upperDecimal;
+    } else {
+      upperBound = Long.parseLong(upper);
+    }
+
+    // Reload the interactive-histogram.html.
+    response.sendRedirect("/interactive-histogram.html");
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    List<Region> regionList = FileParser.getRegionList("../smaps-full.txt");
+
+    // Parse histogram data
+    List<Object[]> histogramData = makeDataArray(regionList);
+
+    // If the user hasn't entered custom bounds yet, use the min and max from regions list.
+    if (!postFired) {
+      setMinMax(regionList);
+    }
+
+    // Construct 2D array to hold the bounds.
+    long[] bounds = new long[2];
+    bounds[0] = lowerBound;
+    bounds[1] = upperBound;
+
+    // Transfer the Java Object arrays into JavaScript Objects (Json).
+    Gson boundsGson = new Gson();
+    Gson histGson = new Gson();
+    String histogramDataJson = histGson.toJson(histogramData);
+    String boundsJson = boundsGson.toJson(bounds);
+
+    // Put the two Json strings into a list to send to histogram.js.
+    List<String> jsonList = new ArrayList<String>();
+    jsonList.add(boundsJson);
+    jsonList.add(histogramDataJson);
+
+    // Write Json to histogram.js
+    response.getWriter().println(jsonList);
+  }
+
+  /** Creates list of 2D Object arrays of data for histogram */
+  static ArrayList<Object[]> makeDataArray(List<Region> regions) {
+    // List must contain both Strings and numbers, so must be of type Object.
+    ArrayList<Object[]> dataArray = new ArrayList<Object[]>();
+
+    // First pair in list must define the format of the data.
+    Object[] labelPair = {"Range", "Size"};
+    dataArray.add(labelPair);
+
+    // Go through the regions and add range/size pairs to the list.
+    for (int i = 0; i < regions.size(); i++) {
+      Region curR = regions.get(i);
+      String range = curR.startLoc() + " - " + curR.endLoc();
+      Object val = (Object) curR.size();
+      Object[] pair = {range, val};
+      dataArray.add(pair);
+    }
+
+    return dataArray;
+  }
+
+  /** Set the two global bounds variables to be the min and max sizes of regions in the list. */
+  static void setMinMax(List<Region> regions) {
+    long min = regions.get(0).size();
```

Comment:
Let's add a check that the size of regions is roughly what you expect/at least none zero.